server/auth/name-space/name-space.service.ts

```diff
@@ -0,0 +1,123 @@
+import { Injectable } from "@nestjs/common";
+import { Model, Types } from "mongoose";
+import { NameSpaceDocument, NameSpace } from "./schemas/name-space.schema";
+import { InjectModel } from "@nestjs/mongoose";
+import slugify from "slugify";
+import { UsersService } from "../../users/users.service";
+import { Roles } from "../../auth/ability/ability.factory";
+
+@Injectable()
+export class NameSpaceService {
+    constructor(
+        @InjectModel(NameSpace.name)
+        private NameSpaceModel: Model<NameSpaceDocument>,
+        private usersService: UsersService
+    ) {}
+
+    listAll() {
+        return this.NameSpaceModel.aggregate([
+            {
+                $lookup: {
+                    from: "users",
+                    localField: "users",
+                    foreignField: "_id",
+                    as: "users",
+                },
+            },
+        ]);
+    }
+
+    async create(nameSpace) {
+        nameSpace.slug = slugify(nameSpace.name, {
+            lower: true,
+            strict: true,
+        });
+
+        nameSpace.users = await this.updateNameSpaceUsers(
+            nameSpace.users,
+            nameSpace.slug
+        );
+
+        return new this.NameSpaceModel(nameSpace).save();
+    }
+
+    async update(id, nameSpaceBody) {
+        const nameSpace = await this.getById(id);
+        const newNameSpace = {
+            ...nameSpace.toObject(),
+            ...nameSpaceBody,
+        };
+
+        newNameSpace.slug = slugify(newNameSpace.name, {
+            lower: true,
+            strict: true,
+        });
+
+        newNameSpace.users = await this.updateNameSpaceUsers(
+            newNameSpace.users,
+            nameSpace.slug
+        );
+        await this.findNameSpaceUsersAndDelete(
+            nameSpace.slug,
+            newNameSpace.users,
+            nameSpace.users
+        );
+
+        return await this.NameSpaceModel.updateOne(
+            { _id: nameSpace._id },
+            newNameSpace
```

Comment:
@pepermao I think this is a good point, we need to sanitise this before doing do update, otherwise it creates a vector to update arbitrary namespaces.